README.md
# Everscale Client

Everscale SDK Client library Python bindings.
Works for Python 3.7+

![PyPI - Python Version](https://img.shields.io/pypi/pyversions/ton-client-py?label=Python)
![PyPI](https://img.shields.io/pypi/v/ton-client-py?label=PyPI)
![PyPI - Downloads](https://img.shields.io/pypi/dm/ton-client-py?label=PyPI%20Downloads)
![GitHub Workflow Status (branch)](https://img.shields.io/github/actions/workflow/status/move-ton/ton-client-py/tonclient-tests.yml?branch=master&label=Test%20MacOS%7CUbuntu%7CWindows%20-%20Python%203.7%7C3.9)
[![Chat on Telegram RU](https://img.shields.io/badge/Chat%20on-Telegram%20RU-blue)](https://t.me/MOVETON_SDK_RU)
[![Chat on Telegram EN](https://img.shields.io/badge/Chat%20on-Telegram%20EN-blue)](https://t.me/MOVETON_SDK_EN)

## Installation

Check if Python 3.7+ is installed

##### MacOS/Linux

```
# Using pipenv
pipenv install ton-client-py

# Using pip
pip install ton-client-py
```

##### Windows

```
# Using pipenv
py -m pipenv install ton-client-py

# Using pip
py -m pip install ton-client-py
```

## Tests

- Clone repository

```
# Clone repository
git clone https://github.com/move-ton/ton-client-py.git

# Go to repo directory
cd ton-client-py
```

- Install dev dependencies

##### MacOS/Linux

```
# Using pipenv
pipenv install --dev

# Using pip
pip install pytest
```

##### Windows

```
# Using pipenv
py -m pipenv install --dev

# Using pip
py -m pip install pytest
```

- Running tests

##### MacOS/Linux

```
# Using pipenv
pipenv run pytest  # Display only module name while testing
pipenv run pytest -v  # Display module and method while testing
pipenv run pytest -v -s --log-cli-level=INFO  # Display methods logging while testing

# Without pipenv
python -m pytest
python -m pytest -v
python -m pytest -v -s --log-cli-level=INFO
```

##### Windows

```
# Using pipenv
py -m pipenv run pytest  # Display only module name while testing
py -m pipenv run pytest -v  # Display module and method while testing
py -m pipenv run pytest -v -s --log-cli-level=INFO  # Display methods logging while testing

# Without pipenv
py -m pytest
py -m pytest -v
py -m pytest -v -s --log-cli-level=INFO
```

- Alternative running tests
  If you have problems with installing `pytest` package you can simply run

```
# For MacOS/Linux
python -m unittest -v

# For Windows
py -m unittest -v
```

## Client

Core client library has sync and async request modes.
Some core methods are available only in async request mode and
this mode is more prefferable, so python client created with async core requests by default.

Create client

```python
from tonclient.types import ClientConfig
from tonclient.client import TonClient

client = TonClient(config=ClientConfig())

# If you need sync core requests for some reason
client_sync_core = TonClient(config=ClientConfig(), is_core_async=False)
```

Client created with default config

```python
from tonclient.types import NetworkConfig, CryptoConfig, AbiConfig, BocConfig, ProofsConfig, ClientConfig


# Default network config is below.
# `None` attributes are filled by core with defaults values:
#     `server_address=''`
#     `endpoints=[]`
#     `network_retries_count=5` (DEPRECATED)
#     `message_retries_count=5`
#     `max_reconnect_timeout=120000`
#     `message_processing_timeout=40000`
#     `wait_for_timeout=40000`
#     `out_of_sync_threshold=15000`
#     `sending_endpoint_count=2`
#     `reconnect_timeout=` (DEPRECATED)
#     `access_key=''`
#     `latency_detection_interval=60000`
#     `max_latency=60000`
#     `query_timeout=60000`
#     `queries_protocol='HTTP'`
network = NetworkConfig(
    server_address='http://localhost', endpoints=None, network_retries_count=None,
    message_retries_count=None, message_processing_timeout=None, reconnect_timeout=None,
    wait_for_timeout=None, out_of_sync_threshold=None, sending_endpoint_count=None,
    access_key=None, max_reconnect_timeout=None, latency_detection_interval=None,
    max_latency=None, query_timeout=None, queries_protocol=None)

# Default crypto config is below.
# `None` attributes are filled by core with defaults values:
#     `mnemonic_dictionary=1`
#     `mnemonic_word_count=12`
#     `hdkey_derivation_path="m/44'/396'/0'/0/0"`
crypto = CryptoConfig(
    mnemonic_dictionary=None, mnemonic_word_count=None, hdkey_derivation_path=None)

# Default abi config is below.
# `None` attributes are filled by core with defaults values:
#     `workchain=0`
#     `message_expiration_timeout=40000`
#     `message_expiration_timeout_grow_factor=1.5`
abi = AbiConfig(
    workchain=None, message_expiration_timeout=None,
    message_expiration_timeout_grow_factor=None)

# Default boc config is below.
# `None` attributes are filled by core with defaults values:
#     `cache_max_size=10000` (10MB)
boc = BocConfig(cache_max_size=None)

# Default proofs config is below.
# `None` attributes are filled by core with defaults values:
#     `cache_in_local_storage=True`
proofs = ProofsConfig(cache_in_local_storage=None)

# Then `ClientConfig` is created
config = ClientConfig(network=network, crypto=crypto, abi=abi, boc=boc, proofs=proofs, local_storage_path=None)
```

You can override initial config while creating a client

```python
from tonclient.types import ClientConfig
from tonclient.client import TonClient, DEVNET_BASE_URLS

config = ClientConfig()
config.network.endpoints = DEVNET_BASE_URLS
config.abi.message_expiration_timeout = 30000

client = TonClient(config=config)
version = client.version()
```

Client contains all core modules and its methods.
You can get full list of modules and methods here:
https://github.com/tonlabs/TON-SDK/blob/master/docs/modules.md
Module method called by template `client.[module].[method]`

```python
from tonclient.types import ClientConfig, ParamsOfParse
from tonclient.client import TonClient, DEVNET_BASE_URLS

config = ClientConfig()
config.network.endpoints = DEVNET_BASE_URLS
client = TonClient(config=config)

# Generate random signing keys
keypair = client.crypto.generate_random_sign_keys()

# Parse account
parse_params = ParamsOfParse(boc='Account base64 BOC')
result = client.boc.parse_account(params=parse_params)
```

You always can get information about method and its arguments in method docstring.

### Methods with callbacks

Some library methods accept `callback` argument to pass additional data to it.
E.g. `net.subscribe_collection`

```python
import time
from datetime import datetime

from tonclient.errors import TonException
from tonclient.types import ClientConfig, ClientError, SubscriptionResponseType, \
    ParamsOfSubscribeCollection, ResultOfSubscription
from tonclient.client import DEVNET_BASE_URLS, TonClient


config = ClientConfig()
config.network.endpoints = DEVNET_BASE_URLS
client = TonClient(config=config)


def __callback(response_data, response_type, loop):
    """
    `loop` in args is just for example.
    It will have value only with `asyncio` and may be replaced by `_` or `*args`
    in synchronous requests
    """
    if response_type == SubscriptionResponseType.OK:
        result = ResultOfSubscription(**response_data)
        results.append(result.result)
    if response_type == SubscriptionResponseType.ERROR:
        raise TonException(error=ClientError(**response_data))

results = []
now = int(datetime.now().timestamp())
q_params = ParamsOfSubscribeCollection(collection='messages', result='created_at', filter={'created_at': {'gt': now}})
subscription = client.net.subscribe_collection(params=q_params, callback=__callback)

while True:
    if len(results) > 0 or int(datetime.now().timestamp()) > now + 10:
        client.net.unsubscribe(params=subscription)
        break
    time.sleep(1)
```

Please, dig in `tonclient/test/test_net.py`, `tonclient/test/test_processing.py`,
`tonclient/test/test_crypto.py`, `tonclient/test/test_debot.py` to get more examples.

## Client and asyncio

```python
from tonclient.types import ClientConfig
from tonclient.client import TonClient, DEVNET_BASE_URLS

# Create client with `is_async=True` argument.
config = ClientConfig()
config.network.endpoints = DEVNET_BASE_URLS
client = TonClient(config=config, is_async=True)

# Get version (simple method with result)
version = await client.version()
```

Please, dig in `tonclient/test/test_async.py` to get more info


CHANGELOG.md
## Version 1.44.3.0

- Binaries updated to `1.44.3`;
- Changes `1.44.2` (https://github.com/tonlabs/ever-sdk/blob/master/CHANGELOG.md#1442--2023-08-22);
- Changes `1.44.3` (https://github.com/tonlabs/ever-sdk/blob/master/CHANGELOG.md#1443--2023-09-05);

## Version 1.44.1.0

- Binaries updated to `1.44.1`;
- Changes `1.43.0` (https://github.com/tonlabs/ever-sdk/blob/master/CHANGELOG.md#1430--2023-05-23);
- Changes `1.43.1` (https://github.com/tonlabs/ever-sdk/blob/master/CHANGELOG.md#1431--2023-06-01);
- Changes `1.43.2` (https://github.com/tonlabs/ever-sdk/blob/master/CHANGELOG.md#1432--2023-06-09);
- Changes `1.43.3` (https://github.com/tonlabs/ever-sdk/blob/master/CHANGELOG.md#1433--2023-06-24);
- Changes `1.44.0` (https://github.com/tonlabs/ever-sdk/blob/master/CHANGELOG.md#1440--2023-07-12);
- Changes `1.44.1` (https://github.com/tonlabs/ever-sdk/blob/master/CHANGELOG.md#1441--2023-07-25);

## Version 1.42.1.0

- Binaries updated to `1.42.1`;
- Changes `1.41.1` (https://github.com/tonlabs/ever-sdk/blob/master/CHANGELOG.md#1411--2023-03-14);
- Changes `1.42.0` (https://github.com/tonlabs/ever-sdk/blob/master/CHANGELOG.md#1420--2023-03-22);
- Changes `1.42.1` (https://github.com/tonlabs/ever-sdk/blob/master/CHANGELOG.md#1421--2023-03-23);

## Version 1.41.0.0

- Binaries updated to `1.41.0`;
- Changes `1.41.0` (https://github.com/tonlabs/ever-sdk/blob/master/CHANGELOG.md#1410--2023-02-13);

## Version 1.40.0.0

- Binaries updated to `1.40.0`;
- Changes `1.40.0` (https://github.com/tonlabs/ever-sdk/blob/master/CHANGELOG.md#1400--2023-01-11);

## Version 1.39.0.0

- Binaries updated to `1.39.0`;
- Changes `1.39.0` (https://github.com/tonlabs/ever-sdk/blob/master/CHANGELOG.md#1390--2022-12-07);

## Version 1.38.1.0

- Binaries updated to `1.38.1`;
- Changes `1.38.0` (https://github.com/tonlabs/ever-sdk/blob/master/CHANGELOG.md#1380--2022-10-06);
- Changes `1.38.1` (https://github.com/tonlabs/ever-sdk/blob/master/CHANGELOG.md#1381--2022-11-10);

## Version 1.37.2.0

- Binaries updated to `1.37.2`;
- Changes `1.37.1` (https://github.com/tonlabs/ever-sdk/blob/master/CHANGELOG.md#1371--2022-08-03);
- Changes `1.37.2` (https://github.com/tonlabs/ever-sdk/blob/master/CHANGELOG.md#1372--2022-08-10);

## Version 1.37.0.0

- Binaries updated to `1.37.0`;
- Changes `1.37.0` (https://github.com/tonlabs/ever-sdk/blob/master/CHANGELOG.md#1370--2022-07-28);

## Version 1.36.1.0

- Binaries updated to `1.36.1`;
- Changes `1.36.1` (https://github.com/tonlabs/ever-sdk/blob/master/CHANGELOG.md#1361--2022-07-18);

## Version 1.36.0.0

- Binaries updated to `1.36.0`;
- Changes `1.36.0` (https://github.com/tonlabs/ever-sdk/blob/master/CHANGELOG.md#1360--2022-07-01);

## Version 1.35.0.0

- Binaries updated to `1.35.0`;
- Changes `1.34.3` (https://github.com/tonlabs/ever-sdk/blob/master/CHANGELOG.md#1343--2022-06-08);
- Changes `1.35.0` (https://github.com/tonlabs/ever-sdk/blob/master/CHANGELOG.md#1350--2022-06-28);

## Version 1.34.2.0

- Binaries updated to `1.34.2`;
- Changes `1.34.0` (https://github.com/tonlabs/ever-sdk/blob/master/CHANGELOG.md#1340--2022-05-18);
- Changes `1.34.1` (https://github.com/tonlabs/ever-sdk/blob/master/CHANGELOG.md#1341--2022-05-26);
- Changes `1.34.2` (https://github.com/tonlabs/ever-sdk/blob/master/CHANGELOG.md#1342--2022-05-30);

**Breaking changes**

- `client.config` changed from instance attribute to method, now should call `client.config()` or `await client.config()`
- `AppDebotBrowser.perform_invoke_debot` method signature changed.
  Now it accepts `config: ClientConfig` instead of `client: TonClient`

## Version 1.33.1.0

- Binaries updated to `1.33.1`;
- Changes `1.33.0` (https://github.com/tonlabs/ever-sdk/blob/master/CHANGELOG.md#1330--2022-05-02);
- Changes `1.33.1` (https://github.com/tonlabs/ever-sdk/blob/master/CHANGELOG.md#1331--2022-05-10);

## Version 1.32.0.1

- Bug fix in `types.AbiData` by [PR#3](https://github.com/move-ton/ton-client-py/pull/3).
  Thanks to [@abionics](https://github.com/abionics)

## Version 1.32.0.0

- Binaries updated to `1.32.0`;
- Changes `1.32.0` (https://github.com/tonlabs/TON-SDK/blob/master/CHANGELOG.md#1320--2022-03-22);

## Version 1.31.0.0

- Binaries updated to `1.31.0`;
- Changes `1.31.0` (https://github.com/tonlabs/TON-SDK/blob/master/CHANGELOG.md#1310--2022-03-09);

## Version 1.30.0.0

- Binaries updated to `1.30.0`;
- Changes `1.27.1` (https://github.com/tonlabs/TON-SDK/blob/master/CHANGELOG.md#1271--2021-12-09);
- Changes `1.28.0` (https://github.com/tonlabs/TON-SDK/blob/master/CHANGELOG.md#1280--2021-12-24);
- Changes `1.28.1` (https://github.com/tonlabs/TON-SDK/blob/master/CHANGELOG.md#1281--2022-01-25);
- Changes `1.29.0` (https://github.com/tonlabs/TON-SDK/blob/master/CHANGELOG.md#1290--2022-02-03);
- Changes `1.30.0` (https://github.com/tonlabs/TON-SDK/blob/master/CHANGELOG.md#1300--2022-02-04);

## Version 1.27.0.1

- Internal refactor: methods' return types suggestion are correct and clear now;

## Version 1.27.0.0

- Binaries updated to `1.27.0`;
- Changes `1.26.0` (https://github.com/tonlabs/TON-SDK/blob/master/CHANGELOG.md#1260--2021-11-25);
- Changes `1.26.1` (https://github.com/tonlabs/TON-SDK/blob/master/CHANGELOG.md#1261--2021-12-01);
- Changes `1.27.0` (https://github.com/tonlabs/TON-SDK/blob/master/CHANGELOG.md#1270--2021-12-03);

## Version 1.25.0.0

- Binaries updated to `1.25.0`;
- Changes `1.25.0` (https://github.com/tonlabs/TON-SDK/blob/master/CHANGELOG.md#1250--2021-11-08);

## Version 1.24.0.0

- Binaries updated to `1.24.0`;
- Changes `1.24.0` (https://github.com/tonlabs/TON-SDK/blob/master/CHANGELOG.md#1240--2021-10-18);

## Version 1.23.0.0

- Binaries updated to `1.23.0`;
- Changes `1.23.0` (https://github.com/tonlabs/TON-SDK/blob/master/CHANGELOG.md#1230--2021-10-05);

## Version 1.22.0.0

tonsdk/lib.py
import ctypes
import os
import json
import logging
import platform
from typing import Dict

from tonsdk.ctypes import InteropString, InteropJsonResponse

logger = logging.getLogger('ton')

BASE_DIR = os.path.dirname(os.path.realpath(__file__))
LIB_VERSION = '0.25.3'
LIB_DIR = os.path.join(BASE_DIR, 'bin')
LIB_FILENAME = f'ton-rust-client-{LIB_VERSION}'

DEVNET_BASE_URL = 'net.ton.dev'
MAINNET_BASE_URL = 'main.ton.dev'

TON_CLIENT_DEFAULT_SETUP = {
    'servers': ['http://localhost'],
    'messageRetriesCount': 1,
    'messageExpirationTimeout': 50000,
    'messageExpirationTimeoutGrowFactor': 1.5,
    'messageProcessingTimeout': 50000,
    'messageProcessingTimeoutGrowFactor': 1.5,
    'waitForTimeout': 30000,
    'accessKey': ''
}


def _get_lib_path():
    plt = platform.system().lower()
    lib_ext_dict = {
        'windows': 'dll',
        'darwin': 'dylib',
        'linux': 'so'
    }
    if plt not in lib_ext_dict:
        raise RuntimeError(
            f'No library for current platform "{plt.capitalize()}"')
    return os.path.join(LIB_DIR, f'{LIB_FILENAME}.{lib_ext_dict[plt]}')


_LIB = ctypes.cdll.LoadLibrary(_get_lib_path())


class TonClient(object):
    TYPE_TEXT = "text"
    TYPE_HEX = "hex"
    TYPE_BASE64 = "base64"

    BOX_OUTPUT_TEXT = "Text"
    BOX_OUTPUT_HEX = "Hex"
    BOX_OUTPUT_HEX_UP = "HexUppercase"
    BOX_OUTPUT_BASE64 = "Base64"

    def __init__(self, **config):
        self.context = _LIB.tc_create_context()
        self.setup({**TON_CLIENT_DEFAULT_SETUP, **config})

    def _request(self, method_name, params=None) -> Dict:
        """
        :param method_name: SDK method name
        :param params: Method params
        :return: Dict
        """
        method_name = InteropString.from_string(method_name)
        params = InteropString.from_string(json.dumps(params or {}))

        _LIB.tc_json_request.restype = ctypes.POINTER(InteropJsonResponse)
        response_ptr = _LIB.tc_json_request(self.context, method_name, params)

        _LIB.tc_read_json_response.restype = InteropJsonResponse
        response = _LIB.tc_read_json_response(response_ptr)
        result = {'success': response.is_success, 'result': response.json}

        _LIB.tc_destroy_json_response(response_ptr)
        return result

    def destroy_context(self):
        return _LIB.tc_destroy_context(self.context)

    def setup(self, settings):
        return self.request(method="setup", params=settings)

    def version(self):
        return self.request(method="version")

    def _str_type_dict(self, string: str, fmt: str) -> Dict:
        """
        Generates dict for API params, based on string format

        :param string: Any string
        :param fmt: One of 'TYPE_x' constants
        :return: Dict
        """
        if fmt not in [self.TYPE_BASE64, self.TYPE_HEX, self.TYPE_TEXT]:
            raise ValueError(
                "One of 'base64, hex, text' should be provided "
                "for 'fmt' param")
        return {fmt: string}

    def request(self, method: str, params=None, raise_exception=True):
        result = self._request(method, params)

        if raise_exception and not result["success"]:
            raise Exception(result["result"])

        return result["result"]

    def mnemonic_generate(self, word_count=24) -> str:
        """
        Generate random mnemonic

        :param word_count:
        :return: str
        """
        params = {"wordCount": word_count}
        return self.request(
            method='crypto.mnemonic.from.random', params=params)

    def mnemonic_from_entropy(
            self, entropy: str, entropy_fmt: str, word_count=24) -> str:
        """
        :param entropy: String as hex, base64 or plain text
        :param entropy_fmt: Entropy string type (TonClient.TYPE_x)
        :param word_count:
        :return: str
        """
        params = {
            "wordCount": word_count,
            "entropy": self._str_type_dict(string=entropy, fmt=entropy_fmt)
        }
        return self.request(
            method="crypto.mnemonic.from.entropy", params=params)

    def mnemonic_verify(self, mnemonic: str) -> bool:
        """
        :param mnemonic:
        :return: bool
        """
        params = {"phrase": mnemonic, "wordCount": len(mnemonic.split(" "))}
        return self.request(method='crypto.mnemonic.verify', params=params)

    def mnemonic_words(self) -> str:
        return self.request("crypto.mnemonic.words")

    def sha512(self, string: str, string_fmt: str) -> str:
        """
        :param string: String as hex, base64 or plain text
        :param string_fmt: String type (TonClient.TYPE_x)
        :return: str
        """
        params = {
            "message": self._str_type_dict(string=string, fmt=string_fmt)
        }
        return self.request(method='crypto.sha512', params=params)

    def sha256(self, string: str, string_fmt: str) -> str:
        """
        :param string: String as hex, base64 or plain text
        :param string_fmt: Entropy string type (TonClient.TYPE_x)
        :return: str
        """
        params = {
            "message": self._str_type_dict(string=string, fmt=string_fmt)
        }
        return self.request(method='crypto.sha256', params=params)

    def scrypt(
            self, data: str, n: int, r: int, p: int, dk_len: int, salt: str,
            salt_fmt: str, password: str, password_fmt: str) -> str:
        """
        :param data: Data to encrypt
        :param n: The CPU/Memory cost parameter. Must be larger than 1,
                a power of 2, and less than 2^(128 * r / 8)
        :param r: The parameter specifies block size
        :param p: The parallelization parameter. Is a positive integer
                less than or equal to ((2^32-1) * 32) / (128 * r)
        :param dk_len: The intended output length. Is the length in octets
                of the key to be derived ("keyLength"); it is a positive
                integer less than or equal to (2^32 - 1) * 32.
        :param salt: Salt string
        :param salt_fmt: Salt string type (TonClient.TYPE_x)
        :param password: Password string
        :param password_fmt: Password string type (TonClient.TYPE_x)
        :return: str
        """
        params = {
            "data": data,
            "salt": self._str_type_dict(string=salt, fmt=salt_fmt),
            "password": self._str_type_dict(string=password, fmt=password_fmt),
            "logN": n,
            "r": r,
            "p": p,
            "dkLen": dk_len
        }
        return self.request(method="crypto.scrypt", params=params)

    def keystore_add(self, keypair: Dict) -> str:
        """
        :param keypair: Keypair dict {"public": str, "secret": str}
        :return: str
        """
        return self.request(method='crypto.keystore.add', params=keypair)

    def keystore_remove(self, index: str):
        """
        :param index: Keystore index to be removed
        :return:
        """
        self.request(method='crypto.keystore.remove', params=str(index))

    def keystore_clear(self):
        """ Clear keystore """
        self.request(method='crypto.keystore.clear')

    def hdkey_xprv_from_mnemonic(self, mnemonic: str) -> str:
        """
        Get BIP32 key from mnemonic

        :param mnemonic:
        :return: str
        """
        params = {"phrase": mnemonic, "wordCount": len(mnemonic.split(" "))}
        return self.request(
            method='crypto.hdkey.xprv.from.mnemonic', params=params)

    def hdkey_xprv_secret(self, bip32_key: str) -> str:
        """
        Get private key from BIP32 key

        :param bip32_key:
        :return: str
        """
        params = {"serialized": bip32_key}
        return self.request(method='crypto.hdkey.xprv.secret', params=params)

    def hdkey_xprv_public(self, bip32_key: str) -> str:
        """
        Get public key from BIP32 key

        :param bip32_key:
        :return: str
        """
        params = {"serialized": bip32_key}
        return self.request(method='crypto.hdkey.xprv.public', params=params)

    def hdkey_xprv_derive_path(self, bip32_key: str, derive_path: str) -> str:
        """
        :param bip32_key:
        :param derive_path:
        :return: str
        """
        params = {"serialized": bip32_key, 'path': derive_path}
        return self.request(
            method='crypto.hdkey.xprv.derive.path', params=params)

    def hdkey_xprv_derive(self, bip32_key: str, index: int) -> str:
        """
        :param bip32_key:
        :param index:
        :return: str
        """
        params = {"serialized": bip32_key, 'index': index}
        return self.request(method='crypto.hdkey.xprv.derive', params=params)

    def factorize(self, number: str) -> Dict:
        """
        :param number:
        :return: Dict
        """
        return self.request(method='crypto.math.factorize', params=number)

    def ton_public_key_string(self, public_key: str) -> str:
        """
        :param public_key:
        :return: str
        """
        return self.request(
            method='crypto.ton_public_key_string', params=public_key)

    def ed25519_keypair(self) -> Dict:
        """ Generate ed25519 keypair """
        return self.request(method='crypto.ed25519.keypair')

    def modular_power(self, base: str, exponent: str, modulus: str) -> str:
        """
        :param base:
        :param exponent:
        :param modulus:
        :return: str
        """
        params = {'base': base, 'exponent': exponent, 'modulus': modulus}
        return self.request(method='crypto.math.modularPower', params=params)

    def nacl_box_keypair(self) -> Dict:
        """ Generate nacl box keypair """
        return self.request(method='crypto.nacl.box.keypair')

    def nacl_sign_keypair(self) -> Dict:
        """ Generate nacl sign keypair """
        return self.request(method='crypto.nacl.sign.keypair')

    def nacl_sign_keypair_from_secret_key(self, secret_key: str) -> Dict:
        """
        Generate nacl sign keypair from secret key

        :param secret_key:
        :return: Dict
        """
        return self.request(
            method='crypto.nacl.sign.keypair.fromSecretKey', params=secret_key)

    def nacl_box_keypair_from_secret_key(self, key: str) -> Dict:
        return self.request(
            method='crypto.nacl.box.keypair.fromSecretKey', params=key)

    def nacl_box(
            self, nonce: str, their_public: str, secret_key: str, message: str,
            message_fmt: str, output_fmt: str = BOX_OUTPUT_HEX) -> str:
        """
        :param nonce:
        :param their_public:
        :param secret_key:
        :param message:
        :param message_fmt:
        :param output_fmt:
        :return: str
        """
        params = {
            "nonce": nonce,
            "theirPublicKey": their_public,
            "secretKey": secret_key,
            "message": self._str_type_dict(string=message, fmt=message_fmt),
            "outputEncoding": output_fmt
        }
        return self.request(method="crypto.nacl.box", params=params)

    def nacl_box_open(
            self, nonce: str, their_public: str, secret_key: str,
            message: str, message_fmt: str, output_fmt: str = BOX_OUTPUT_TEXT)\
            -> str:
        """
        :param nonce:
        :param their_public:
        :param secret_key:
        :param message:
        :param message_fmt:
        :param output_fmt:
        :return: str
        """
        params = {
            "nonce": nonce,
            "theirPublicKey": their_public,
            "secretKey": secret_key,
            "message": self._str_type_dict(string=message, fmt=message_fmt),
            "outputEncoding": output_fmt
        }
        return self.request(method="crypto.nacl.box.open", params=params)

    def nacl_sign(
            self, key: str, message: str, message_fmt: str,
            output_fmt: str = BOX_OUTPUT_HEX) -> str:
        """
        :param key:
        :param message:
        :param message_fmt:
        :param output_fmt:
        :return: str
        """
        params = {
            "key": key,
            "message": self._str_type_dict(string=message, fmt=message_fmt),
            "outputEncoding": output_fmt
        }
        return self.request(method='crypto.nacl.sign', params=params)

    def nacl_sign_detached(
            self, key: str, message: str, message_fmt: str,
            output_fmt: str = BOX_OUTPUT_HEX) -> str:
        """
        :param key:
        :param message:
        :param message_fmt:
        :param output_fmt:
        :return: str
        """
        params = {
            "key": key,
            "message": self._str_type_dict(string=message, fmt=message_fmt),
            "outputEncoding": output_fmt
        }
        return self.request(method="crypto.nacl.sign.detached", params=params)

    def nacl_sign_open(
            self, key: str, message: str, message_fmt: str,
            output_fmt: str = BOX_OUTPUT_HEX) -> str:
        """
        :param key:
        :param message:
        :param message_fmt:
        :param output_fmt:
        :return: str
        """
        params = {
            "key": key,
            "message": self._str_type_dict(string=message, fmt=message_fmt),
            "outputEncoding": output_fmt
        }
        return self.request(method="crypto.nacl.sign.open", params=params)

    def nacl_secret_box(
            self, nonce: str, their_public: str, message: str,
            message_fmt: str, output_fmt: str = BOX_OUTPUT_HEX) -> str:
        """
        :param nonce:
        :param their_public:
        :param message:
        :param message_fmt:
        :param output_fmt:
        :return: str
        """
        params = {
            "nonce": nonce,
            "key": their_public,
            "message": self._str_type_dict(string=message, fmt=message_fmt),
            "outputEncoding": output_fmt
        }
        return self.request(method="crypto.nacl.secret.box", params=params)

    def nacl_secret_box_open(
            self, nonce: str, their_public: str, secret_key: str,
            message: str, message_fmt: str, output_fmt: str = BOX_OUTPUT_TEXT)\
            -> str:
        """
        :param nonce:
        :param their_public:
        :param secret_key:
        :param message:
        :param message_fmt:
        :param output_fmt:
        :return: str
        """
        params = {
            "nonce": nonce,
            "key": their_public,
            "secretKey": secret_key,
            "message": self._str_type_dict(string=message, fmt=message_fmt),
            "outputEncoding": output_fmt
        }
        return self.request(
            method="crypto.nacl.secret.box.open", params=params)

    def query(
            self, table: str, filter: Dict, result: str, order: Dict = None,
            limit: int = None) -> any:
        """
        :param table:
        :param filter: {"field_name": {"rule": "condition"}}
        :param result:
        :param order: {"path": str (field name), "direction": str (ASC|DESC)}
        :param limit:
        :return:
        """
        params = {
            "table": table,
            "filter": json.dumps(filter),
            "result": result,
            "order": order,
            "limit": limit
        }
        result = self.request(method="queries.query", params=params)

        return result["result"]

    def query_wait_for(
            self, table: str, filter: Dict, result: str, order: Dict = None,
            limit: int = None, timeout: int = None) -> any:
        """
        :param table:
        :param filter:
        :param result:
        :param order:
        :param limit:
        :param timeout:
        :return:
        """
        params = {
            "table": table,
            "filter": json.dumps(filter),
            "result": result,
            "order": order,
            "limit": limit,
            "timeout": timeout
        }
        result = self.request(method="queries.wait.for", params=params)

        return result["result"]

    def query_subscribe(
            self, table: str, filter: Dict, result: str) -> int:
        """
        :param table:
        :param filter:
        :param result:
        :return: Handle index
        """
        params = {
            "table": table,
            "filter": json.dumps(filter),
            "result": result
        }
        result = self.request(method="queries.subscribe", params=params)

        return result["handle"]

    def query_get_next(self, handle: int) -> any:
        """
        :param handle:
        :return:
        """
        params = {"handle": handle}
        result = self.request(method="queries.get.next", params=params)

        return result["result"]

    def query_unsubscribe(self, handle: int) -> None:
        """
        :param handle:
        :return:
        """
        params = {"handle": handle}
        return self.request(method="queries.unsubscribe", params=params)

    # async def _request_async(self, method_name, params: Dict, req_id: int,
    #                          cb: Callable = _on_result):
    #     logger.debug('Create request (async)')
    #     logger.debug(f'Context: {self.context}')
    #
    #     method = method_name.encode()
    #     method_interop = InteropString(
    #         ctypes.cast(method, ctypes.c_char_p), len(method))
    #     logger.debug(f'Fn name: {method}')
    #
    #     params = json.dumps(params).encode()
    #     params_interop = InteropString(
    #         ctypes.cast(params, ctypes.c_char_p), len(params)
    #     )
    #     logger.debug(f'Data: {params}')
    #
    #     on_result = OnResult(cb)
    #     response = _LIB.tc_json_request_async(
    #         self.context, method_interop, params_interop,
    #         ctypes.c_int32(req_id), on_result)
    #     logger.debug(f'Response: {response}')
    #
    #     _LIB.tc_destroy_json_response(response)
    #     return response


# def _on_result(request_id: int, result_json: InteropString,
#                error_json: InteropString, flags: int):
#     """ Python callback for lib async request """
#     logger.debug('Async callback fired')
#     logger.debug(
#         f'Request ID: {request_id}\n'
#         f'Result JSON: {result_json}\n'
#         f'Error JSON: {error_json}\n'
#         f'Flags: {flags}\n')
#
#     if result_json.len > 0:
#         logger.debug('Result JSON: ', result_json.content)
#     elif error_json.len > 0:
#         logger.debug('Error JSON: ', error_json.content)
#     else:
#         logger.debug('No response data')
